src/components/scroll-tracker.cjsx

```diff
@@ -0,0 +1,121 @@
+
+ScrollTrackerParentMixin =
+
+  getInitialState: ->
+    scrollPoints: []
+    scrollState: {}
+    scrollTopBuffer: 0
+
+  setScrollTopBuffer: ->
+    scrollTopBuffer = GetPositionMixin.getTopPosition(@getDOMNode())
+    @setState({scrollTopBuffer})
+
+  setScrollPoint: (scrollPoint, scrollState) ->
+    scrollPointData = _.extend({scrollPoint: scrollPoint}, scrollState)
+    @state.scrollPoints.push(scrollPointData)
+    @sortScrollPoints()
+
+  unsetScrollPoint: (unsetScrollPoint) ->
+    @state.scrollPoints = _.reject @state.scrollPoints, (scrollPoint) ->
+      scrollPoint.scrollPoint is unsetScrollPoint
+    @sortScrollPoints()
+
+  sortScrollPoints: ->
+    sortedDescScrollPoints = _.sortBy @state.scrollPoints, (scrollData) ->
+      -1 * scrollData.scrollPoint
+
+    @setState({scrollPoints: sortedDescScrollPoints})
+
+  getScrollStateByScroll: (scrollTop) ->
+    scrollState = _.find @state.scrollPoints, (scrollData) =>
+      scrollTop > (scrollData.scrollPoint - @state.scrollTopBuffer - 2)
+
+    scrollState or _.last(@state.scrollPoints)
+
+  areKeysSame: (key, keyToCompare) ->
+    key is keyToCompare or parseInt(key) is parseInt(keyToCompare)
+
+  getScrollStateByKey: (stepKey) ->
+    scrollStateIndex = _.findLastIndex @state.scrollPoints, (scrollData) =>
+      @areKeysSame(scrollData.key, stepKey)
+
+    @state.scrollPoints[scrollStateIndex]
+
+  setScrollState: ->
+    scrollState = @getScrollStateByScroll(@state.scrollTop)
+    @setState({scrollState})
+
+    @props.setScrollState(scrollState)
+
+  isScrollPointsStable: (compareState) ->
+    _.isEqual(@state.scrollPoints, compareState.scrollPoints)
+
+  shouldCheckForScrollingState: (state) ->
+    state ?= @state
+    not _.isEmpty(state.scrollPoints) and not _.isUndefined(state.scrollState) and @isScrollPointsStable(state)
+
+  componentDidMount: ->
+    @setScrollTopBuffer()
+    @scrollToKey(@props.currentStep) if @props.currentStep?
+
+  componentWillUpdate: (nextProps, nextState) ->
+    return unless @shouldCheckForScrollingState(nextState)
+    willScrollStateKeyChange = not @areKeysSame(nextState.scrollState.key, @state.scrollState.key)
+    @props.goToStep(nextState.scrollState.key) if willScrollStateKeyChange
+
+  componentDidUpdate: (prevProps, prevState) ->
+    return unless @shouldCheckForScrollingState(prevState)
+    doesScrollStateMatch = @areKeysSame(prevState.scrollState.key, @getScrollStateByScroll(@state.scrollTop)?.key)
+    didCurrentStepChange = not @areKeysSame(@props.currentStep, prevState.scrollState?.key)
+
+    unless doesScrollStateMatch
+      @setScrollState()
+    else if didCurrentStepChange
+      @scrollToKey(@props.currentStep)
+
+  scrollToKey: (stepKey) ->
+    return unless stepKey?
+    scrollState = @getScrollStateByKey(stepKey)
+    return unless scrollState?
+    window.scrollTo(0, (scrollState?.scrollPoint - @state.scrollTopBuffer))
+
+module.exports = {ScrollTrackerMixin, ScrollTracker, ScrollTrackerParentMixin}
```

Comment:
Imma gonna trust this works.  I don't see anything wrong with it, it's just complex and makes me a bit nervous that it may have subtle bugs that are missed because it takes awhile to figure it out.  But I don't really have anything that comes to mind immediately to on how best to simplify it either.